File: src/model/typed/TypedSessionProvider.ts
/**
 * Copyright © 2023 Nevis Security AG. All rights reserved.
 */

import { TypedData } from './TypedData';
import { CookieContainer } from '../../authorization/CookieContainer';
import {
	CookieSessionProvider,
	JwtSessionProvider,
	SessionProvider,
} from '../../authorization/SessionProvider';

class SessionProviderData {
	cookies?: Array<CookieContainer>;
	jwt?: string;

	constructor(cookies?: Array<CookieContainer>, jwt?: string) {
		this.cookies = cookies;
		this.jwt = jwt;
	}
}

export abstract class TypedSessionProvider extends TypedData<SessionProviderData> {
	abstract data: SessionProviderData;

	static create(sessionProvider: SessionProvider): TypedSessionProvider {
		if (sessionProvider instanceof CookieSessionProvider) {
			return new TypedCookieSessionProvider(sessionProvider as CookieSessionProvider);
		} else if (sessionProvider instanceof JwtSessionProvider) {
			return new TypedJwtSessionProvider(sessionProvider as JwtSessionProvider);
		} else {
			throw new Error(`Unknown session provider (${sessionProvider.constructor.name}).`);
		}
	}
}

export class TypedCookieSessionProvider extends TypedSessionProvider {
	type = 'CookieSessionProvider';
	data: SessionProviderData;

	constructor(wrapped: CookieSessionProvider) {
		super();
		this.data = new SessionProviderData(wrapped.cookies);
	}
}

export class TypedJwtSessionProvider extends TypedSessionProvider {
	type = 'JwtSessionProvider';
	data: SessionProviderData;

	constructor(wrapped: JwtSessionProvider) {
		super();
		this.data = new SessionProviderData(undefined, wrapped.jwt);
	}
}

File: ios/Model/Typed/TypedSessionProvider.swift
//
// Copyright © 2023 Nevis Security AG. All rights reserved.
//

import NevisMobileAuthentication

struct TypedSessionProvider: Codable {
	// MARK: Properties

	let wrapped: SessionProvider

	// MARK: Initialization

	init(wrapped: SessionProvider) {
		self.wrapped = wrapped
	}

	// MARK: Coding Keys

	enum SessionProviderType: String, Encodable {
		case cookie = "CookieSessionProvider"
		case jwt = "JwtSessionProvider"
	}
	
	enum CodingKeys: String, CodingKey {
		case cookies
		case jwt
	}

	// MARK: Decodable

	init(from decoder: Decoder) throws {
		let container = try decoder.container(keyedBy: TypedCodingKeys.self)
		let type = try container.decode(String.self, forKey: .type)
		switch type {
		case SessionProviderType.cookie.rawValue:
			wrapped = try container.decode(CookieSessionProvider.self, forKey: .data)
		case SessionProviderType.jwt.rawValue:
			wrapped = try container.decode(JwtSessionProvider.self, forKey: .data)
		default:
			throw DecodingError.dataCorrupted(
				DecodingError.Context(
					codingPath: container.codingPath,
					debugDescription: "Failed to decode session provider!"
				)
			)
		}
	}
	
	// MARK: Encodable

	func encode(to encoder: Encoder) throws {
		var container = encoder.container(keyedBy: TypedCodingKeys.self)
		switch wrapped {
		case is CookieSessionProvider:
			try container.encode(SessionProviderType.cookie, forKey: .type)
		case is JwtSessionProvider:
			try container.encode(SessionProviderType.jwt, forKey: .type)
		default:
			throw EncodingError.invalidValue(
				wrapped,
				EncodingError.Context(
					codingPath: container.codingPath,
					debugDescription: "Failed to encode session provider!"
				)
			)
		}

		try wrapped.encode(to: container.superEncoder(forKey: .data))
	}
}
